File: c/example_game_loop.c
#include "idport.h"
#define FPS 30

int main(void)
{
  idp_api_t mp;
  idp_response_t *response = NULL;
  const char *user_id = "test_user", *stream_id = "1";
  float probability;
  int err;

  /* To start using IDport, we setup an API with with a brain-computer
   * interface server address. */
  idp_init(&mp, "http://localhost:5000");


  /* Start the game loop. */
  for(int frame=0;; ++frame) {
    /* Four times a second, we request the server to detect important
     * patterns in the incoming EEG data. To keep this example simple,
     * we don't keep a list with with parallel transfers --- which we
     * recommended for a real game in order to reliably support frequent
     * (> 8 Hz) update rates. */
    if (frame % (FPS / 4) == 0) {
      response = idp_get_detection(&mp, user_id, stream_id);
    }

    /* We hate interrupting your the game. So we update our active
     * requests, without stalling the game: */
    idp_update(&mp);

    /* If we have received  a completed responses from the server, we
     * can use the detected brain state. Here we use a fake detector
     * that gives random result for testing purposes. */
    switch (response->status) {
    case IDP_RESP_READY:
      err = idp_read_detection(response, "random", &probability);
      idp_response_destroy(response); /* Free slot for new requests. */
      if (err) {
        fprintf(stderr, "Could not decode message, code = %d!\n", err);
        exit(-1);
      }

      if (probability > 0.80) {
        printf("Detected random event with high confidence: p = %.2f.\n",
               probability);
      }
      break;

    case IDP_RESP_INVALID:
      fprintf(stderr, "Cleaning up invalid response.\n");
      idp_response_destroy(response); /* Free slot for new requests. */
      break;

    default:
      break;
    }

    printf("."); fflush(stdout); /* Show smooth progress. */
    usleep(1e6/FPS);
  }

  idp_destroy(&mp);
  return 0;
}

File: python/example_gameloop.py
import time, argparse, sys
import idport

# Right now, everything is working synchronously. It would be quite easy
# to turn this in to async networking using hooks in request: 
# [1] http://docs.python-requests.org/en/latest/user/advanced/#event-hooks

FPS = 30

if __name__ == '__main__':
  # Add a convenient CLI interface.
  parser = argparse.ArgumentParser(description='Game-loop example.')
  parser.add_argument('--url', default='http://localhost:5000',
    help='URL of IDport REST API')
  parser.add_argument('--user_id', default='test-user')
  parser.add_argument('--stream_id', default='test-stream')

  args = parser.parse_args()

  frame = 0
  while True:
    try:
      if frame % 8 == 0:
        # We request a new detection. For now, this happens in a
        # blocking fashion.
        d = idport.get_detection(args.url, args.user_id, args.stream_id)
        if d['random'] > .9:
          print ('Detected random event with high probability: p=%.2f .' % 
            d['random']),

      #if frame % 100 == 0:
      #  print 'Sending an annotation...'
      #  print idport.post_annotation(args.url, args.user_id, args.stream_id,
      #    'python-example', 'Hi!')

    except NotImplementedError as e:
      print 'Error:', e
    
    sys.stdout.write('.'); sys.stdout.flush()  # show progress.

    time.sleep(1./FPS)
    frame += 1
    
